[Safer CPP] Address issues in CSSFontSelector and CSSFontStyleRangeValue

roberto-apple opened the pull request:
#### 642dd9043e5d7ced8a8fe77de6c3445650b14316
<pre>
[Safer CPP] Address issues in CSSFontSelector and CSSFontStyleRangeValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=295682">https://bugs.webkit.org/show_bug.cgi?id=295682</a>
<a href="https://rdar.apple.com/155481948">rdar://155481948</a>

Reviewed by Ryosuke Niwa.

Address Safer CPP issues in CSSFontSelector and CSSFontStyleRangeValue.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::~CSSFontSelector):
(WebCore::CSSFontSelector::fontFaceSet):
(WebCore::CSSFontSelector::buildStarted):
(WebCore::CSSFontSelector::buildCompleted):
(WebCore::CSSFontSelector::addFontFaceRule):
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::resolveGenericFamily):
(WebCore::CSSFontSelector::fontRangesForFamily):
(WebCore::CSSFontSelector::fallbackFontCount):
(WebCore::CSSFontSelector::fallbackFontAt):
(WebCore::CSSFontSelector::isSimpleFontSelectorForDescription const):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
(WebCore::CSSFontStyleRangeValue::customCSSText const):
* Source/WebCore/css/CSSFontStyleRangeValue.h:

Canonical link: <a href="https://commits.webkit.org/297356@main">https://commits.webkit.org/297356@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fdf61863ef5a9ee577b3ac4aca7abb387f6b3dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31132 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117498 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61734 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84715 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25401 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100343 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65163 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18478 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61320 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94787 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120721 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38515 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28623 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93637 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38891 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93463 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23813 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38573 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16351 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34552 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38404 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43881 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38069 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39771 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -394,9 +393,8 @@ FontRanges CSSFontSelector::fontRangesForFamily(const FontDescription& fontDescr
-    auto* document = dynamicDowncast<Document>(m_context.get());
-    auto* face = m_cssFontFaceSet->fontFace(fontDescriptionForLookup->fontSelectionRequest(), familyForLookup);
-    if (face) {
+    CheckedPtr document = dynamicDowncast<Document>(m_context.get());
```
Why not use RefPtr instead?

roberto-apple replied:
The issue was classified as "Unchecked raw pointer or reference not provably backed by checked variable" in the scan results; I understood it as requiring a Checked type. Same for the other two comments. Should I still change each to RefPtr?

rniwa replied:
Yes, please use `RefPtr` instead.

rniwa commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -433,11 +431,12 @@ RefPtr<Font> CSSFontSelector::fallbackFontAt(const FontDescription& fontDescript
-    if (auto* document = dynamicDowncast<Document>(m_context.get()); document && document->settingsValues().webAPIStatisticsEnabled)
+    auto& pictographFontFamily = context->settingsValues().fontGenericFamilies.pictographFontFamily();
+    RefPtr font = FontCache::forCurrentThread()->fontForFamily(fontDescription, pictographFontFamily);
+    if (CheckedPtr document = dynamicDowncast<Document>(context.get()); document && document->settingsValues().webAPIStatisticsEnabled)
```
Ditto. We typically use RefPtr for Node subclasses.

rniwa commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -450,7 +449,7 @@ bool CSSFontSelector::isSimpleFontSelectorForDescription() const
 
     // FIXME: remove this when we fix counter style rules mutation.
-    if (auto* document = dynamicDowncast<Document>(m_context.get())) {
+    if (CheckedPtr document = dynamicDowncast<Document>(m_context.get())) {
```
Ditto.

roberto-apple committed:
[Safer CPP] Address issues in CSSFontSelector and CSSFontStyleRangeValue
https://bugs.webkit.org/show_bug.cgi?id=295682
rdar://155481948

Reviewed by Ryosuke Niwa.

Address Safer CPP issues in CSSFontSelector and CSSFontStyleRangeValue.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::~CSSFontSelector):
(WebCore::CSSFontSelector::fontFaceSet):
(WebCore::CSSFontSelector::buildStarted):
(WebCore::CSSFontSelector::buildCompleted):
(WebCore::CSSFontSelector::addFontFaceRule):
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::resolveGenericFamily):
(WebCore::CSSFontSelector::fontRangesForFamily):
(WebCore::CSSFontSelector::fallbackFontCount):
(WebCore::CSSFontSelector::fallbackFontAt):
(WebCore::CSSFontSelector::isSimpleFontSelectorForDescription const):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
(WebCore::CSSFontStyleRangeValue::customCSSText const):
* Source/WebCore/css/CSSFontStyleRangeValue.h:

Canonical link: https://commits.webkit.org/297356@main
